SYCL/ESIMD/api/simd_memory_access.cpp

```diff
@@ -0,0 +1,181 @@
+    A = new T[size];
+  } else {
+    A = reinterpret_cast<T *>(sycl::malloc_shared(size, q));
+  }
+
+  for (unsigned i = 0; i < size; ++i) {
+    A[i] = 0;
+  }
+
+  try {
+    if constexpr (IsAcc) {
+      buffer<T, 1> buf(A, range<1>(size));
+      range<1> glob_range{size / N};
+
+      auto e = q.submit([&](handler &cgh) {
+        auto acc = buf.template get_access<access::mode::read_write>(cgh);
+        Kernel<T, N, true> kernel(acc);
+        cgh.parallel_for(glob_range, kernel);
+      });
+    } else {
+      range<1> glob_range{size / N};
+
+      auto e = q.submit([&](handler &cgh) {
+        Kernel<T, N, false> kernel(A);
+        cgh.parallel_for(glob_range, kernel);
+      });
+    }
+    q.wait_and_throw();
+  } catch (cl::sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free_mem<IsAcc>(A, q);
+    return e.get_cl_code();
+  }
+
+  int err_cnt = 0;
+
+  for (unsigned i = 0; i < size; ++i) {
+    T gold = (T)i;
+    T val = A[i];
+
+    if (val != gold) {
+      if (++err_cnt < 10) {
+        using T1 = typename char_to_int<T>::type;
+        std::cout << "failed at index " << i << ": " << (T1)val
+                  << " != " << (T1)gold << " (gold)\n";
+      }
+    }
+  }
+  if (err_cnt > 0) {
+    std::cout << "  pass rate: "
+              << ((float)(size - err_cnt) / (float)size) * 100.0f << "% ("
+              << (size - err_cnt) << "/" << size << ")\n";
+  }
+
+  free_mem<IsAcc>(A, q);
+
+  std::cout << (err_cnt > 0 ? "  FAILED\n" : "  Passed\n");
+  return err_cnt > 0 ? false : true;
+}
+
+int main(int argc, char **argv) {
+  size_t size = 32 * 7;
+
+  if (argc > 1) {
+    size = atoi(argv[1]);
+    size = size == 0 ? 128 : size;
+  }
+  if (size % 32 != 0) {
+    std::cerr << "*** ERROR: size (" << size << ") must be a multiple of 32\n";
+    return 2;
+  }
+  std::cout << "Using size=" << size << "\n";
+  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
+
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+  bool passed = true;
+  passed &= test<char, 32, true>(q, size);
+  passed &= test<unsigned char, 16, true>(q, size);
+  passed &= test<short, 32, true>(q, size);
+  passed &= test<short, 16, true>(q, size);
+  passed &= test<unsigned short, 8, true>(q, size);
+  passed &= test<int, 32, true>(q, size);
+  passed &= test<unsigned int, 32, true>(q, size);
+  passed &= test<float, 32, true>(q, size);
+
+  passed &= test<char, 32, false>(q, size);
+  passed &= test<unsigned char, 16, false>(q, size);
+  passed &= test<short, 32, false>(q, size);
+  passed &= test<short, 16, false>(q, size);
+  passed &= test<unsigned short, 8, false>(q, size);
+  passed &= test<int, 32, false>(q, size);
+  passed &= test<unsigned int, 32, false>(q, size);
+  passed &= test<float, 32, false>(q, size);
+
+  std::cout << (passed ? "=== Test passed\n" : "=== Test FAILED\n");
+  return passed ? 0 : 1;
+}
```

Comment:
nit: would it become more readable if we just split USM- and accessor-based tests into separate files? Some `if constexpr` would go away.

Reply:
Yes, it might. But I'd prefer to have this in a single file to avoid code duplication and reduce maintenance burden.